// rword/dist/rword.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generate_indexes_1 = require("./lib/generate-indexes");
const shuffle_words_1 = require("./lib/shuffle-words");
const english_1 = require("./words/english");
class rword {
    static get words() {
        return english_1.words;
    }
    /** Randomly generates words from the words array. */
    static generate(count = 1, opt) {
        opt = Object.assign({
            contains: /.*/,
            length: '3-10',
            capitalize: 'none'
        }, opt);
        let length = {};
        const contains = typeof opt.contains == 'string' ? new RegExp(opt.contains) : opt.contains;
        // Convert opt.length to an object
        if (typeof opt.length == 'string' && opt.length.indexOf('-') > -1) {
            const l = opt.length.split('-');
            length = { start: +l[0], end: +l[1] };
        }
        // Convert number or string number ('5') to an object
        else if (typeof opt.length != 'object') {
            length = { exactly: +opt.length };
        }
        let pool = [];
        // Skip filtering if possible
        if (contains.toString() == '/.*/' &&
            length.start == 3 &&
            length.end == 10) {
            pool = english_1.words;
        }
        // Filter out unwanted words
        else {
            pool = english_1.words.filter(word => {
                // Filter out words that don't match length
                if (length.exactly) {
                    if (word.length != length.exactly)
                        return false;
                }
                else {
                    if (word.length < length.start || word.length > length.end)
                        return false;
                }
                // Filter out words that don't contain regex
                if (contains)
                    return contains.test(word);
                else
                    return true;
            });
        }
        // No matches
        if (!pool.length)
            return count == 1 ? '' : [];
        // Generate indexes for words to return
        const indexes = generate_indexes_1.generateIndexes(pool.length, count);
        const temp = [];
        // Select words by index
        indexes.forEach(index => temp.push(pool[index]));
        pool = temp;
        // Capitalize words
        switch (opt.capitalize) {
            case 'all':
                pool = pool.map(w => w.toUpperCase());
                break;
            case 'first':
                pool = pool.map(w => w[0].toUpperCase() + w.slice(1));
                break;
        }
        // Returns string or array of strings
        return count == 1 ? pool[0] : pool;
    }
    /** Shuffles words and globalPool arrays. */
    static shuffle() {
        shuffle_words_1.shuffleWords(english_1.words);
        shuffle_words_1.shuffleWords(rword.globalPool);
    }
    /**
     * A simple generator that pulls words from a prefilled global pool. Should
     *  be preferred over `rword.generate()` if custom filters are not needed as
     *  this method can in certain instances be many times faster.
     * @param count - Words to return. Throws error if greater than `10`
     */
    static generateFromPool(count = 1) {
        if (count > 10)
            throw 'Too many words requested. Use rword.generate().';
        // Fill globalPool
        if (count > rword.globalPool.length)
            rword.globalPool = this.generate(500);
        const pool = rword.globalPool.splice(0, count);
        return count == 1 ? pool[0] : pool;
    }
}
rword.globalPool = [];
exports.rword = rword;
// Populate words[] for the first time
if (!english_1.words.length)
    rword.shuffle();
